test/unit/math/prim/functor/hcubature_test.cpp

```diff
@@ -0,0 +1,197 @@
+#include <stan/math/prim/functor.hpp>
+#include <stan/math/prim/fun.hpp>
+
+#include <gtest/gtest.h>
+#include <vector>
+
+namespace hcubature_test {
+
+struct my_params {
+  long double x;
+  double y;
+};
+
+template <typename T_x, typename T_p>
+stan::return_type_t<T_x, T_p> f1(const T_x& x, const T_p& p) {
+  using T_x_ref = stan::ref_type_t<T_x>;
+  T_x_ref x_ref = x;
+  stan::scalar_seq_view<T_x_ref> x_vec(x_ref);
+  return cos(x_vec[0]);
+}
+
+template <typename T_x, typename T_p>
+stan::return_type_t<T_x, T_p> f2(const T_x& x, const T_p& p) {
+  using T_x_ref = stan::ref_type_t<T_x>;
+  T_x_ref x_ref = x;
+  stan::scalar_seq_view<T_x_ref> x_vec(x_ref);
+  return cos(x_vec[0]) * cos(x_vec[1]);
+}
+
+template <typename T_x, typename T_p>
+stan::return_type_t<T_x, T_p> f3(const T_x& x, const T_p& p) {
+  using T_x_ref = stan::ref_type_t<T_x>;
+  T_x_ref x_ref = x;
+  stan::scalar_seq_view<T_x_ref> x_vec(x_ref);
+  my_params* pars = static_cast<my_params*>(p);
+  long double radius = (pars->x);
+  double result;
+  if (std::pow(x_vec[0], 2) + std::pow(x_vec[1], 2) < std::pow(radius, 2)) {
+    result = 1;
+  } else {
+    result = 0;
+  }
+  return result;
+}
+
+template <typename T_x, typename T_p>
+stan::return_type_t<T_x, T_p> f4(const T_x& x, const T_p& p) {
+  using T_x_ref = stan::ref_type_t<T_x>;
+  using namespace stan::math;
+  T_x_ref x_ref = x;
+  stan::scalar_seq_view<T_x_ref> x_vec(x_ref);
+  my_params* pars = static_cast<my_params*>(p);
+  double a = (pars->y);
+  double s = std::pow(x_vec[0] - 0.5, 2) + std::pow(x_vec[1] - 0.5, 2);
```

Comment:
Can you use more verbose names throughout the code please?

Reply:
Comes with next commit.